zarr/core.py

```diff
@@ -1948,6 +1969,25 @@ def _process_for_setitem(self, ckey, chunk_selection, value, fields=None):
             else:
                 chunk[chunk_selection] = value
 
+        # clear chunk if it only contains the fill value
+        if not self._write_empty_chunks:
+            if self.dtype == 'object':
+                # we have to flatten the result of np.equal to handle outputs like
+                # [np.array([True,True]), True, True]
+                is_empty = all(flatten(np.equal(chunk, self.fill_value, dtype='object')))
+            else:
+                is_empty = np.all(chunk == self._fill_value)
+
+            if is_empty:
```

Comment:
Since these lines are repeated lots of times, @d-v-b, do you want to make a function or two? e.g. `def clear_if_empty(check_obj_dtype=False) -> bool:`, returns True if the block was found to be empty and successfully cleared.

Reply:
Noted! In https://github.com/zarr-developers/zarr-python/pull/738/commits/044a9b8c64a37d3ba9bfe247ad23114bd3411611 I functionalize approximately as per your suggestion